Animal feeding improvements

Pr0methean opened the pull request:
- Implements use of food for taming, horse healing, growth, or to set Love Mode.
- Animals won't consume food when it has none of these effects.
- `GlowAbstractHorse.canBreed()` can no longer return true if untamed.
- Implements the duration limit and the particle effect for Love Mode.

Pr0methean committed:
Animal feeding fixes:
* Implements horse healing.
* Implements growing.
* Food no longer consumed if !canBreed() && !canGrow() (except by horses, who are healed by food).

Pr0methean committed:
Implements love mode duration and particle emission.

Pr0methean committed:
Bug fix

Pr0methean committed:
Update GlowAbstractHorseTest

Pr0methean committed:
Update more test cases

Pr0methean committed:
Implement use of food for taming

Pr0methean committed:
Override tryFeed instead of entityInteract in GlowHorse

Pr0methean committed:
Merge remote-tracking branch 'origin/dev' into animalFeeding

Pr0methean committed:
Update testSetBreedTrueBaby

aramperes commented:
What kind of live testing is required for this? Could you describe some testing steps?

Pr0methean commented:
Test that:

- Animals descending from GlowAbstractHorse eventually become tamed when fed repeatedly
- Baby animals grow when fed
- Injured animals heal when fed
- Adult animals, except untamed GlowAbstractHorse, enter Love Mode when fed (currently doesn't affect behavior, but should create heart particles)
- More than one of those things can happen with the same piece of food
- Food isn't consumed if none of those things can happen
- Love Mode eventually expires

FlorentClarret commented:
I tried your test and actually noticed some stuff:
- Heart particles are indeed created when an animal is fed, but the particles appear on their feet instead of above them, which looks a bit strange. 
- Food is still consumed even if nothing else happens. I think it's because the canBreed method return always true if the animal is adult. I tested on cows and pigs.

The other things seem to work, I will also try to take a look soon to improve animal feeding

Pr0methean commented:
@FlorentClarret For your test where food was falsely consumed, would the cow and pig have been expected to enter Love Mode? If not, why not?

FlorentClarret commented:
I just gave wheat to a cow, which was not in love mode and the cow entered in love mode as expected, but after that the wheat is still consumed if I keep giving it wheat. I tried on vanilla an it's not the case, the wheat is not consumed if the cow is already in love mode.

Pr0methean committed:
Changes from review:
- Heart particles should now spawn above the mob's head.
- Food won't be consumed to refresh love mode if already active.

Pr0methean commented:
Both issues should be fixed now.

Pr0methean committed:
Refactor

Pr0methean committed:
Refactor

Pr0methean committed:
Bug fix: tamed horse won't eat when at full health and not breedable

Pr0methean committed:
First unit test of feeding

Pr0methean committed:
Override for horses: need taming

Pr0methean committed:
Test bug fix

Pr0methean committed:
Another test case: no love mode if recently bred

Pr0methean committed:
Override test: mules cannot breed

Pr0methean committed:
Explanatory comment

FlorentClarret reviewed (approved):
Looks good to me now. Thanks @Pr0methean

aramperes commented on src/main/java/net/glowstone/entity/GlowAnimal.java:
```diff
@@ -44,6 +44,30 @@ protected int getAmbientDelay() {
+     * @return true if the item should be consumed; false otherwise
+     */
+    protected boolean tryFeed(Material type, GlowPlayer player) {
+        if (getBreedingFoods().contains(type)) {
```
Remove a nesting level by inverting this if-statement.

Pr0methean replied:
Done.

aramperes commented on src/main/java/net/glowstone/entity/GlowAnimal.java:
```diff
@@ -44,6 +44,30 @@ protected int getAmbientDelay() {
+     */
+    protected boolean tryFeed(Material type, GlowPlayer player) {
+        if (getBreedingFoods().contains(type)) {
+            if (canBreed() && (getInLove() <= 0)) {
```
The parentheses for the second condition are unnecessary.

Pr0methean replied:
Removed, although I would've thought they'd improve readability (since one would guess that `getInLove()` returned boolean without knowing otherwise).

aramperes commented on src/main/java/net/glowstone/entity/GlowAnimal.java:
```diff
@@ -44,6 +44,30 @@ protected int getAmbientDelay() {
+                setInLove(1000); // TODO get the correct duration
+                player.incrementStatistic(Statistic.ANIMALS_BRED);
+                return true;
+            } else {
```
The else-statement is unnecessary.

boeotian replied:
Good code is a balance of function & readability.

Pr0methean replied:
Removing it in this case doesn't hurt readability.

aramperes commented on src/main/java/net/glowstone/entity/passive/GlowAbstractHorse.java:
```diff
@@ -35,6 +36,19 @@
 
+    @Override
+    protected boolean tryFeed(Material food, GlowPlayer player) {
+        if (isAdult() && !isTamed()) {
```
Remove a nesting level by inverting this if-statement.

Pr0methean replied:
Done.

aramperes commented on src/main/java/net/glowstone/entity/passive/GlowAbstractHorse.java:
```diff
@@ -104,6 +123,11 @@ private int getHorseFlags() {
     }
 
+    protected int computeTamingAmount(Material material) {
+        // TODO
```
Please document what's the intended functionality of the method.

Pr0methean replied:
Done, and renamed to `computeDomestication` to match the Bukkit nomenclature. Parameter renamed to `food`.

Pr0methean committed:
Refactors

aramperes reviewed (approved):
Good work! Sorry for the delay.
